services/datamanager/datasync/sync.go

```diff
@@ -124,45 +120,47 @@ func NewManager(logger golog.Logger, partID string, client v1.DataSyncServiceCli
 func (s *syncer) Close() {
 	s.closed.Store(true)
 	s.cancelFunc()
-	if s.conn != nil {
-		if err := s.conn.Close(); err != nil {
-			s.logger.Errorw("error closing datasync server connection", "error", err)
-		}
-	}
 	s.backgroundWorkers.Wait()
 	close(s.syncErrs)
 	s.logRoutine.Wait()
 	//nolint:errcheck
 	_ = s.logger.Sync()
+	if s.conn != nil {
+		if err := s.conn.Close(); err != nil {
+			s.logger.Errorw("error closing datasync server connection", "error", err)
+		}
+	}
 }
 
-func (s *syncer) SyncDirectory(dir string) {
-	paths := getAllFilesToSync(dir, s.lastModifiedMillis)
-	for _, p := range paths {
-		newP := p
-		s.backgroundWorkers.Add(1)
-		goutils.PanicCapturingGo(func() {
-			defer s.backgroundWorkers.Done()
-			select {
-			case <-s.cancelCtx.Done():
+func (s *syncer) SyncFile(path string) {
+	started := make(chan struct{})
+	s.backgroundWorkers.Add(1)
+	goutils.PanicCapturingGo(func() {
+		started <- struct{}{}
+		defer s.backgroundWorkers.Done()
+		select {
+		case <-s.cancelCtx.Done():
+			return
+		default:
+			if !s.markInProgress(path) {
 				return
-			default:
-				if !s.markInProgress(newP) {
-					return
-				}
-				//nolint:gosec
-				f, err := os.Open(newP)
-				if err != nil {
-					// Don't log if the file does not exist, because that means it was successfully synced and deleted
-					// in between paths being built and this executing.
-					if !errors.Is(err, os.ErrNotExist) {
-						s.syncErrs <- errors.Wrap(err, "error opening file")
-					}
-					return
+			}
+			//nolint:gosec
+			f, err := os.Open(path)
+			if err != nil {
+				// Don't log if the file does not exist, because that means it was successfully synced and deleted
+				// in between paths being built and this executing.
+				if !errors.Is(err, os.ErrNotExist) {
+					s.logger.Errorw("error opening file", "error", err)
 				}
+				return
+			}
 
-				if datacapture.IsDataCaptureFile(f) {
-					captureFile, err := datacapture.ReadFile(f)
+			if datacapture.IsDataCaptureFile(f) {
```

Comment:
I know this is outside the scope of this PR but why do we have a separate data capture file type?

Reply:
We need some standardized way of storing these SensorData. Storing a single reading per file would be (generally) very inefficient, so we store the sensor readings as a series of length delimited proto structs with:
- A leading metadata message
- N SensorData

The custom file type is basically to allow us to abstract that away and not deal directly with file pointers, and only deal with reading/writing readings to disk using a handful of convenience methods like ReadNext/WriteNext/ReadMetadata/etc. 

It's all [in this file](https://github.com/viamrobotics/rdk/blob/3196954084a416675ee8cc88dc462035f93f9cce/services/datamanager/datacapture/data_capture_file.go).

Reply:
Thank you for explaining! :)